Generate release notes from commit messages

erkarl opened the pull request:
https://github.com/lightninglabs/lightning-app/issues/480

This PR will add a `generate-release-notes` script that will generate `CHANGELOG.md` based on new commits since the latest tag.

I didn't add automatic tag publishing because we might want to edit the release notes manually before publishing.

tanx commented:
@erkarl thanks for the PR. The main issue is updating the Github Release notes text via api. Perhaps there's a curl command that can be called with a `GH_TOKEN` env var?

erkarl commented:
Gotcha, I'll see what I can do about that.

erkarl commented:
@tanx I've updated the script to hit the GitHub releases endpoint to create a new draft release and include the commit messages as release notes.

![screen shot 2018-08-21 at 16 29 35](https://user-images.githubusercontent.com/1056569/44404677-1a362680-a560-11e8-8f77-4869495de8b4.png)

Usage: `GH_TOKEN=123 ./release.sh v2.0`

`GH_TOKEN` can be generated from [github.com/settings/tokens](https://github.com/settings/tokens) and has to have `Full control of private repositories` permission.

tanx commented on release.sh:
```diff
@@ -0,0 +1,41 @@
+    echo Current release: `sed -n 's/.*"version": "\(.*\)".*/\1/p' ${PACKAGE}`
+    exit 1
+fi
+RELEASE=$1
```
I don't fully understand this block. Should the version provided in the package.json be enough to identify the corresponding github release?

tanx replied:
> No, new version name must be provided as an argument. If you don't provide one the script will exit printing you the current version name for convenience.

Just to clarify, I run `npm version 0.2.2-alpha` locally and then push the git tag and updates package.json to the master. So the package.json already has to new version.

erkarl replied:
Oh, thanks for clearing this up. I'll make the necessary changes. The script will then just take the version from `package.json` and generate a draft release notes to Github.

tanx replied:
Cool. Sounds good. Thanks

tanx commented on release.sh:
```diff
@@ -0,0 +1,41 @@
+curl -H "Content-Type: application/json" \
+  -H "Authorization: token $GH_TOKEN" \
+  --data @$TEMP_DATA_JSON \
+  $GITHUB_API_ENDPOINT
```
Cool 👍

tanx commented on release-notes.sh:
```diff
@@ -0,0 +1,29 @@
+CURRENT_TAG=v${RELEASE_NAME}
+PREVIOUS_TAG=$(git describe --tags --abbrev=0 $(git rev-list --tags --skip=1 --max-count=1))
+RELEASE_NOTES="$(git log ${PREVIOUS_TAG}..HEAD --merges --pretty='* %b [%h](http://github.com/lightninglabs/lightning-app/commit/%H)\n')"
+DATA_JSON='{"tag_name": "'"$CURRENT_TAG"'", "name": "'"$RELEASE_NAME"'", "body": "'"$RELEASE_NOTES"'", "prerelease": true, "draft": true}'
```
FYI we don't publish as `prerelease` anymore.

erkarl commented:
> Have you tried what happens when there is already a draft for the release there? Because the binaries should already have been pushed to the same draft before this new script get executed.

@tanx it looks like it currently creates a new draft with the same name. Ideally I'd like it to append to the existing draft.

erkarl commented:
@tanx I've updated the release-notes script to update an existing draft release generated by Travis. I got fed up with bash and refactored the script in JavaScript. Didn't use NodeJS `https` API because I already had the curl commands.

Usage: `node release-notes.js`

![screen shot 2018-09-14 at 16 27 49](https://user-images.githubusercontent.com/1056569/45553657-2c447580-b83d-11e8-9e2b-e0e67f1c00ec.png)

tanx commented:
> @tanx I've updated the release-notes script to update an existing draft release generated by Travis. I got fed up with bash and refactored the script in JavaScript. Didn't use NodeJS `https` API because I already had the curl commands.

Ok. Just curious about the choice of node since the code calls shell commands via node now. Did you run into some issue with bash?

erkarl commented:
@tanx In hindsight the shell script is probably a little easier to read. I've reverted back the changes and added the update existing release functionality to the shell script. For comparison the JS version is [available here](https://github.com/erkarl/lightning-app/blob/release-notes-js/release-notes.js).
